contracts/multisig-prover/src/execute.rs

```diff
@@ -126,31 +128,116 @@ pub fn update_worker_set(deps: DepsMut, env: Env) -> Result<Response, ContractEr
+
+            let key_gen_msg = multisig::msg::ExecuteMsg::KeyGen {
+                key_id,
+                snapshot,
+                pub_keys_by_address: participants
+                    .into_iter()
+                    .map(|(participant, pub_key)| {
+                        (
+                            participant.address.to_string(),
+                            (KeyType::Ecdsa, pub_key.as_ref().into()),
+                        )
+                    })
+                    .collect(),
+            };
+
+            Ok(Response::new().add_message(wasm_execute(config.multisig, &key_gen_msg, vec![])?))
+        }
+        Some(cur_worker_set) => {
+            if !should_update_worker_set(
+                &new_worker_set,
+                &cur_worker_set,
+                config.worker_set_diff_threshold as usize,
+            ) {
+                return Err(ContractError::WorkerSetUnchanged);
+            }
+
+            NEXT_WORKER_SET.save(deps.storage, &(new_worker_set.clone(), snapshot))?;
+            let mut builder = CommandBatchBuilder::new(config.destination_chain_id);
+            builder.add_new_worker_set(new_worker_set)?;
+
+            let batch = builder.build()?;
+
+            COMMANDS_BATCH.save(deps.storage, &batch.id, &batch)?;
+            REPLY_BATCH.save(deps.storage, &batch.id)?;
+
+            let start_sig_msg = multisig::msg::ExecuteMsg::StartSigningSession {
+                key_id: cur_worker_set.id(), // TODO remove the key_id
+                msg: batch.msg_to_sign(),
+            };
+
+            Ok(Response::new().add_submessage(SubMsg::reply_on_success(
+                wasm_execute(config.multisig, &start_sig_msg, vec![])?,
+                START_MULTISIG_REPLY_ID,
+            )))
+        }
     }
+}
+
+pub fn confirm_worker_set(deps: DepsMut) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
 
-    NEXT_WORKER_SET.save(deps.storage, &new_worker_set)?;
-    let mut builder = CommandBatchBuilder::new(config.destination_chain_id);
-    builder.add_new_worker_set(new_worker_set)?;
+    let (worker_set, snapshot) = NEXT_WORKER_SET.load(deps.storage)?;
 
-    let batch = builder.build()?;
+    let query = voting_verifier::msg::QueryMsg::IsWorkerSetConfirmed {
+        new_operators: worker_set.clone().try_into()?,
+    };
 
-    let start_sig_msg = multisig::msg::ExecuteMsg::StartSigningSession {
-        key_id: "static".to_string(), // TODO remove the key_id
-        msg: batch.msg_to_sign(),
+    let is_confirmed: bool = deps.querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
+        contract_addr: config.voting_verifier.to_string(),
+        msg: to_binary(&query)?,
+    }))?;
+
+    if !is_confirmed {
+        return Err(ContractError::WorkerSetNotConfirmed);
+    }
+
+    CURRENT_WORKER_SET.save(deps.storage, &worker_set)?;
+    NEXT_WORKER_SET.remove(deps.storage);
+    KEY_ID.save(deps.storage, &worker_set.hash().to_hex())?;
```

Comment:
ok this is easier to comprehend, but why using that `.id()` method above then?

Reply:
same question

Reply:
changed to `.id()` since key_id will be removed in another PR to use worker set id instead. Then `.id()` will make more sense